### PR TITLE
Add CBOR/CWT encoding

### LINKED ISSUE #3: Document the CWT variation of the status list
As described in the title, detail the CWT representation of the status list which is complementary to the JWT representation.

### LINKED ISSUE #109: Use reference to RFC9052 instead of RFC8152
Use reference to RFC9052 instead of RFC8152

### LINKED ISSUE #110: Rewrite abstract, intro and diagram to use JOSE/COSE centric
The abstract, intro, diagrams wrt referenced tokens description should refer to COSE and JOSE instead of CWT/JWT. Otherwise, non-CWT COSE objects such as ISO 18013-5 MSO cannot use this spec. Same applies to SD-JWT VC encoded secured using JWS.

### PR DESCRIPTION
**Pull request by awoie**
Closes #3
Closes #109 
Closes #110 

## 📑 Description
This PR adds:
- CBOR/CWT encoding for status list and status list info
- Example implementation for CBOR/CWT

Not included yet: examples for referenced tokens (COSE, CWT) but that should be trivial to add.

## Preview Link

[click here for rendered preview of PR](https://vcstuff.github.io/draft-ietf-oauth-status-list/awoie/add-cwt/draft-ietf-oauth-status-list.html)

### CODE REVIEW

**Comment by paulbastian** [comment]
The decoded cbor is:
`{"bits": 1, "lst": h'78dadbb918000217015d'}`

The list does not seem to be valid zlib.

**Comment by c2bo** [comment]
I've tried with cyberchef and it seems to work for me:
`0x78dadbb918000217015d` -> zlib inflate = `b9 a3`

**Comment by paulbastian** [comment]
You are right, I probably tried on the hexstring :D

**Comment by awoie** [comment]
I was wondering whether it would be better to define a CDDL for Status, StatusList and StatusListInfo. Basically, having a high-level CDDl description of all objects and then having some specific sections for JWT, COSE, CWT that define how they are used in the corresponding objects.

**Comment by awoie** [comment]
Also keep in mind that `status` claim for CWT will require IANA to assign us an integer number, so we can use it in the CWT.

File: draft-ietf-oauth-status-list.md [hunk 1]
```diff
@@ -192,12 +192,28 @@ This section defines the structure for a JSON-encoded Status List:
...
-{::include ./examples/status_list_encoding}
+{::include ./examples/status_list_encoding_json}
 ~~~~~~~~~~
 
 ## Status List in CBOR Format {#status-list-cbor}
 
-TBD
+This section defines the `StatusList` structure for a CBOR-encoded Status List.
+
+The `StatusList` structure is a map (Major Type 5) and defines the following entries:
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:202]: Are we considering to register a CBOR tag for StatusList itself?

Comment (awoie) [reply draft-ietf-oauth-status-list.md:202]: No, this is not needed.

**Review by paulbastian — CHANGES_REQUESTED** [review]

File: draft-ietf-oauth-status-list.md [hunk 2]
```diff
@@ -239,7 +255,43 @@ The following is a non-normative example for a Status List Token in JWT format:
...
+
+2. The CWT MUST be digitally signed using an asymmetric cryptographic algorithm. Relying parties MUST reject the JWT if it is using a Message Authentication Code (MAC) algorithm. Relying parties MUST reject CWTs with an invalid signature.
+
+3. Relying parties MUST reject CWTs that are not valid in all other respects per "CBOR Web Token (CWT)" {{RFC8392}}.
+
+4. Application of additional restrictions and policy are at the discretion of the verifying party.
+
+The following is a non-normative example for a Status List Token in CWT format:
+
+~~~~~~~~~~
```
[the remark this answers is not in the document]
Comment (awoie) [reply draft-ietf-oauth-status-list.md:286]: Yep, the typ header is not defined in IANA yet. There is draft for that though. I can add the typ header but the label (14) is not confirmed yet.

File: draft-ietf-oauth-status-list.md [hunk 3]
```diff
@@ -239,7 +255,43 @@ The following is a non-normative example for a Status List Token in JWT format:
 
 ## Status List Token in CWT Format {#status-list-token-cwt}
 
-TBD
+The Status List Token MUST be encoded as a "CBOR Web Token (CWT)" according to {{RFC8392}}.
+
+Applications MAY use the media type `statuslist+cwt` for a CWT in accordance with the rules outlined in this specification.
```
[the remark this answers is not in the document]
Comment (awoie) [reply draft-ietf-oauth-status-list.md:260]: This is what CWT says about the media type. As noted above typ header in CWT/COSE is pending.

File: draft-ietf-oauth-status-list.md [hunk 4]
```diff
@@ -56,14 +56,14 @@ informative:
...
 The status list data structures themselves are also represented as JWTs or CWTs.
 
 --- middle
 
 # Introduction
 
-JSON Web Tokens (JWTs) {{RFC7519}} and CBOR Web Tokens (CWTs) {{RFC8392}} as secure token formats, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time, which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.
+JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Object Signing and Encryption (COSE) {{RFC8152}} objects, and CBOR Web Tokens (CWTs) {{RFC8392}} as secure token formats, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time, which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.
```
[the remark this answers is not in the document]
Comment (c2bo) [reply draft-ietf-oauth-status-list.md]: That sounds like a good idea imho

**Comment by c2bo** [comment]
There is a problem with the width of the CBOR diagnostic notation examples. The datatracker only allows a width of 72, but cborg defaults to 100 and apparently can't be configured in CLI mode.

**Comment by awoie** [comment]
> There is a problem with the width of the CBOR diagnostic notation examples. The datatracker only allows a width of 72, but cborg defaults to 100 and apparently can't be configured in CLI mode.

I can write a small JS script that uses the cborg library and configures the width to 72. Would this be even possible? I didn't check by myself.

**Comment by c2bo** [comment]
Yes that would be possible and is the only option I have found so far. I was searching for a pure python solution (which would be cleaner imho) but that seems to be not possible right now.

File: draft-ietf-oauth-status-list.md (hunk 1 above)
Comment (paulbastian) [reply draft-ietf-oauth-status-list.md:202]: Why not? The StatusList is a separate data structure that may live on its own, dependent from the StatusListToken. Doesn't it make sense in this case or should this be solved by media types alone?

Comment (awoie) [reply draft-ietf-oauth-status-list.md:202]: Tags are usually not needed. In all cases, the structure of the CBOR is pre-defined. That is why CBOR tags are not really needed in most cases.

File: draft-ietf-oauth-status-list.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (awoie) [reply draft-ietf-oauth-status-list.md:286]: Content type is about the secured payload which could be statuslist+cbor. typ could be statulist+cwt. Both can be used.

File: draft-ietf-oauth-status-list.md [hunk 5]
```diff
@@ -56,14 +56,14 @@ informative:
 
 --- abstract
 
-This specification defines status list data structures for representing the status of JSON Web Tokens (JWTs) {{RFC7519}} and CBOR Web Tokens (CWTs) {{RFC8392}}.
+This specification defines status list data structures for representing the status of JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Web Tokens (CWTs) {{RFC8392}}, and other tokens secured by CBOR Object Signing and Encryption {{RFC9052}}.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:59]:
```suggestion
This specification defines status list data structures for representing the status of JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Web Tokens (CWTs) {{RFC8392}}, and other tokens secured by CBOR Object Signing and Encryption {{RFC9052}}.
This specification defines status list data structures and processing rules for representing the status of tokens secured by JSON Object Signing and Encryption {{IANA.JOSE}} or CBOR Object Signing and Encryption {{RFC9052}}, such as JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Web Tokens (CWTs) {{RFC8392}} and ISO mdoc.
```

File: draft-ietf-oauth-status-list.md [hunk 6]
```diff
@@ -56,14 +56,14 @@ informative:
...
 The status list data structures themselves are also represented as JWTs or CWTs.
 
 --- middle
 
 # Introduction
 
-JSON Web Tokens (JWTs) {{RFC7519}} and CBOR Web Tokens (CWTs) {{RFC8392}} as secure token formats, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time, which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.
+JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Object Signing and Encryption (COSE) {{RFC9052}} objects, and CBOR Web Tokens (CWTs) {{RFC8392}} as secure token formats, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time, which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:66]:
```suggestion
Token formats secured by JOSE {{IANA.JOSE}} or COSE {{RFC9052}, such as JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Web Tokens (CWTs) {{RFC8392}} and ISO mdoc, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time, which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.
```

File: draft-ietf-oauth-status-list.md [hunk 7]
```diff
@@ -32,52 +32,73 @@ normative:
...
     title: "JSON Web Token Claims"
     target: "https://www.iana.org/assignments/jwt/jwt.xhtml"
+  IANA.CWT:
+    author:
+      org: "IANA"
+    title: "CBOR Web Token (CWT) Claims"
+    target: "https://www.iana.org/assignments/cwt/cwt.xhtml"
+  CWT.typ:
+    author: 
+      fullname: "M. B. Jones"
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:63]:
```suggestion
      name: "M. B. Jones"
```

File: draft-ietf-oauth-status-list.md [hunk 8]
```diff
@@ -239,7 +276,41 @@ The following is a non-normative example for a Status List Token in JWT format:
...
+
+1. The CWT MAY contain other claims.
+
+2. The CWT MUST be digitally signed using an asymmetric cryptographic algorithm. Relying parties MUST reject the CWT if it is using a Message Authentication Code (MAC) algorithm. Relying parties MUST reject CWTs with an invalid signature.
+
+3. Relying parties MUST reject CWTs that are not valid in all other respects per "CBOR Web Token (CWT)" {{RFC8392}}.
+
+4. Application of additional restrictions and policy are at the discretion of the verifying party.
+
+The following is a non-normative example for a Status List Token in CWT format:
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:303]:
```suggestion
The following is a non-normative example for a Status List Token in CWT format (not including the type header yet):
```

File: draft-ietf-oauth-status-list.md [hunk 9]
```diff
@@ -281,9 +352,37 @@ The following is a non-normative example for a decoded header and payload of a R
...
-## Referenced Token in CWT/CBOR Format {#referenced-token-cwt}
+## Referenced Token in CWT Format {#referenced-token-cwt}
 
-TBD
+The Referenced Token MUST be encoded as a "COSE Web Token (CWT)" object according to {{RFC8392}}.
+
+The following content applies to the JWT Claims Set:
+
+* `1` (issuer): REQUIRED. Same definition as `iss` claim in [](#referenced-token-jwt).
+* `65535` (status): REQUIRED. The status claim is encoded as a `Status` CBOR structure and MUST include at least one data item that refers to a status mechanism. Each data item in the `Status` CBOR structure comprises a key-value pair, where the key must be a CBOR text string (Major Type 3) specifying the identifier of the status mechanism, and the corresponding value defines its contents. This specification defines the following data items:
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:362]: recheck what pre-assigned number to get

File: draft-ietf-oauth-status-list.md [hunk 10]
```diff
@@ -281,9 +352,37 @@ The following is a non-normative example for a decoded header and payload of a R
 }
 ~~~
 
-## Referenced Token in CWT/CBOR Format {#referenced-token-cwt}
+## Referenced Token in CWT Format {#referenced-token-cwt}
 
-TBD
+The Referenced Token MUST be encoded as a "COSE Web Token (CWT)" object according to {{RFC8392}}.
+
+The following content applies to the JWT Claims Set:
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:359]:
```suggestion
The following content applies to the CWT Claims Set:
```

File: draft-ietf-oauth-status-list.md [hunk 11]
```diff
@@ -610,6 +761,7 @@ for their valuable contributions, discussions and feedback to this specification
 * relax requirements on referenced token
 * clarify Deflate / zlib compression
 * make a reference to the Issuer-Holder-Verifier model of SD-JWT VC
+* add COSE/CWT/BOR encoding
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:764]:
```suggestion
* add COSE/CWT/CBOR encoding
```

**Review by paulbastian — APPROVED** [review]

**Review by c2bo — APPROVED** [review]

**Review by tplooker — APPROVED** [review]

**Pull request merged**